### DISCUSSION TITLE
[Mailer] Command mailer:test does not work with SSL

### DISCUSSION THREAD

**Discussion by PyroShape** [symfony/symfony#50001] [Q&A] [answered]
### Symfony version(s) affected

6.2.8

### Description

I am trying to send email with 3rd party transport (MailJet).
I would like to test that the config is good before going any further.

I do `php bin/console mailer:test email.adress@something.com`
The console give me back this error message :
`Connection could not be established with host "ssl://in-v3.mailjet.com:465": stream_socket_client(): SSL operation failed with code 1. 
OpenSSL Error messages: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed`

### How to reproduce

The 3rd party transport MailJet is instaled with `composer require symfony/mailjet-mailer`

Inside the .env file I add this : `MAILER_DSN_MAIN=mailjet+smtp://blabla_ACESS_KEY:blabla_SECRET_KEY@default`

Inside the config mailer.yaml file :
`framework:
    mailer:
        transports:
            main: '%env(MAILER_DSN_MAIN)%'`

I also enabled the TLS support for the symfony server with : `symfony server:ca:install`

**Comment by PhilETaylor** [comment]
I was unable to replicate this.

1. Signed up for MailJet, got my API key/secret from https://app.mailjet.com/account/apikeys
2. `symfony new symfony49990 --version="6.2.*" --webapp`
3. `composer require symfony/mailjet-mailer`
4. Configured `MAILER_DSN=mailjet+smtp://REDACTED:REDACTED@api.mailjet.com` in .env
5. `php bin/console mailer:test phil@phil-taylor.com -vvv`

<img width="1400" alt="ScreenShot-2023-04-11-09 42 00" src="https://user-images.githubusercontent.com/400092/231105365-3290b166-fdc5-4c73-90f4-14497a1b21b4.png">

(Received email from Mallet saying I had not set up this test sender, which proves they received my smtp message) 

<img width="993" alt="ScreenShot-2023-04-11-09 43 55" src="https://user-images.githubusercontent.com/400092/231105854-f300fd87-0edc-4c78-b285-b4c8cf438b2b.png">


6. Changed MAILER_DSN to use API as a test 

<img width="1103" alt="ScreenShot-2023-04-11-09 42 36" src="https://user-images.githubusercontent.com/400092/231105549-c26e10a3-ad6c-406e-9a40-b34825667ea7.png">

**Comment by PyroShape** [comment]
I have tried on a new fresh project as you did :

1. With smtp with the recipe configuration : `MAILER_DSN=mailjet+smtp://ACCESS_KEY:SECRET_KEY@in-v3.mailjet.com` => still same message
2. With smtp with your host configuration : `MAILER_DSN=mailjet+smtp://ACCESS_KEY:SECRET_KEY@api.mailjet.com` => still same message
3. With api with the default host like in the documentation : `MAILER_DSN=mailjet+api://ACCESS_KEY:SECRET_KEY@default` => it works !


Two topics here :

1. SMTP still doesn't work for me (and I have to find by myself why because you can't reproduce)
2. The documentation and the recipe configuration are different : the documentation said `mailjet+smtp://ACCESS_KEY:SECRET_KEY@default` or `mailjet+api://ACCESS_KEY:SECRET_KEY@default`, and the recipe configuration is `mailjet+smtp://PUBLIC_KEY:PRIVATE_KEY@in-v3.mailjet.com` or `mailjet+api://PUBLIC_KEY:PRIVATE_KEY@api.mailjet.com`

**Comment by PhilETaylor** [comment]
To put the API connection to bed - that will work as its a HTTP call and not a SMTP call - different service, thus different SSL.

https://github.com/symfony/symfony/blob/9de1fd76b5509d31e3b99b112c7ad21a7eb32171/src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetApiTransport.php#L72

The DSN `@default` is actually hard coded and is `in-v3.mailjet.com` as per 

https://github.com/symfony/symfony/blob/9de1fd76b5509d31e3b99b112c7ad21a7eb32171/src/Symfony/Component/Mailer/Bridge/Mailjet/Transport/MailjetSmtpTransport.php#L22

In the interests of trying to understand what's happening, At the terminal command line what is your output for these commands please. 

```
dig a in-v3.mailjet.com
dig aaaa in-v3.mailjet.com
openssl s_client -connect in-v3.mailjet.com:465 -4
openssl s_client -connect in-v3.mailjet.com:465 -6
openssl s_client -starttls smtp -connect in-v3.mailjet.com:465 -4
openssl s_client -starttls smtp -crlf -connect in-v3.mailjet.com:465 -6
```

**Comment by PyroShape** [comment]
I don't have dig command on my windows computer so I can only test openssl.

1. `openssl s_client -connect in-v3.mailjet.com:465 -4`
[tests_SMTP_1.txt](https://github.com/symfony/symfony/files/11211030/tests_SMTP_1.txt)
2. `openssl s_client -connect in-v3.mailjet.com:465 -6`
[tests_SMTP_2.txt](https://github.com/symfony/symfony/files/11211031/tests_SMTP_2.txt)
3. `openssl s_client -starttls smtp -connect in-v3.mailjet.com:465 -4`
[tests_SMTP_3.txt](https://github.com/symfony/symfony/files/11211032/tests_SMTP_3.txt)
4. `openssl s_client -starttls smtp -crlf -connect in-v3.mailjet.com:465 -6`
[tests_SMTP_4.txt](https://github.com/symfony/symfony/files/11211033/tests_SMTP_4.txt)

**Comment by PhilETaylor** [comment, marked as answer]
And there is the root of your problem... Avast Antivirus.

Your choice of Antivirus software on your computer... which is intercepting outgoing connections - with an invalid self-signed certificate... 

```
depth=1 OU = generated by Avast Antivirus for SSL/TLS scanning, O = Avast Web/Mail Shield, CN = Avast Web/Mail Shield Root
verify error:num=19:self signed certificate in certificate chain
verify return:1
depth=1 OU = generated by Avast Antivirus for SSL/TLS scanning, O = Avast Web/Mail Shield, CN = Avast Web/Mail Shield Root
verify return:1
```

**Reply by PyroShape** [reply]
Arf :/
So why is it working with the api throught https request ?

**Reply by PhilETaylor** [reply]
I dont run windows or Avast so I could not tell you how they proxy those requests sorry. 

try running `openssl s_client -connect in-v3.mailjet.com:443` which is the SSL port web api would use and see what that shows.

**Reply by PyroShape** [reply]
That shows that avast doesn't analyse this port.
I have find that it analyse the outgoing SMTP so I disabled it and mailer:test command works perfectly now.
Thank you Phil !

**Comment by PhilETaylor** [comment]
Also as "Avast is able to detect and decrypt TLS/SSL protected traffic" some security conscious people (like me) would call that a MITM exploit... they are literally intercepting your outgoing secure traffic, decrypting it, looking at it, and then sending it onward.... gulp.... I wonder if their app also does that with SSL traffic from your web browser when you are logging into your bank's website... gulp.
